src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets

```diff
@@ -40,4 +40,90 @@
+  -->
+
+  <Target Name="InitializeDeterministicSourceRootPaths"
+          DependsOnTargets="$(InitializeSourceInformation)"
+          Condition="'$(DeterministicSourcePaths)' == 'true'">
+
+    <ItemGroup Condition="'@(_MappedSourceRoot)' != ''">
+      <_MappedSourceRoot Remove="@(_MappedSourceRoot)" />
+    </ItemGroup>
+
+    <MapSourceRoots SourceRoots="@(SourceRoot)" Condition="'$(DeterministicSourcePaths)' == 'true'">
+      <Output TaskParameter="MappedSourceRoots" ItemName="_MappedSourceRoot" />
+    </MapSourceRoots>
+
+    <ItemGroup Condition="'$(DeterministicSourcePaths)' == 'true'">
+      <SourceRoot Remove="@(SourceRoot)" />
+      <SourceRoot Include="@(_MappedSourceRoot)" />
+    </ItemGroup>
+  </Target>
+
+  <!--
+    Unless specified otherwise enable deterministic source root (PathMap) when building deterministically on CI server, but not for local builds.
+    In order for the debugger to find source files when debugging a locally built binary the PDB must contain original, unmapped local paths.
+    
+    Backward compatibility:
+    Capture the value of DeterministicSourcePaths set by the project before it possibly changes to true on CI server.
+    If DeterministicSourcePaths is NOT set to true by the project we do not report an error when SourceRoot is not available to avoid breaking existing projects,
+    otherwise we report an error since PathMap can't be calculated without SourceRoot and the build would not be deterministic.
+  -->
+  <PropertyGroup>
+    <_DeterministicSourcePathsOriginal>$(DeterministicSourcePaths)</_DeterministicSourcePathsOriginal>
+    <DeterministicSourcePaths Condition="'$(DeterministicSourcePaths)' == '' and '$(Deterministic)' == 'true' and '$(ContinuousIntegrationBuild)' == 'true'">true</DeterministicSourcePaths>
+  </PropertyGroup>
+
+  <!-- 
+    =======
+    PathMap
+    =======
+
+    If DeterministicSourcePaths is true sets PathMap based on SourceRoots.
+
+    This target requires SourceRoot to be initialized in order to calculate the PathMap.
+    If SourceRoot doesn't contain any non-nested roots an error is reported if DeterministicSourcePaths has been explicitly set to true by the project.
+
+  -->
+
+  <Target Name="_SetPathMapFromSourceRoots"
+          DependsOnTargets="InitializeDeterministicSourceRootPaths"
+          BeforeTargets="CoreCompile"
+          Condition="'$(DeterministicSourcePaths)' == 'true'">
+
+    <ItemGroup>
+      <_TopLevelSourceRoot Include="@(SourceRoot)" Condition="'%(SourceRoot.NestedRoot)' == ''"/>
+    </ItemGroup>
+
+    <Error Text="SourceRoot must contain top-level source roots when DeterministicSourcePaths is true"
+           Condition="'@(_TopLevelSourceRoot)' == '' and '$(_DeterministicSourcePathsOriginal)' == 'true'" />
+
+    <PropertyGroup Condition="'@(_TopLevelSourceRoot)' != ''">
+      <!-- TODO: Report error/warning if /pathmap doesn't cover all emitted source paths: https://github.com/dotnet/roslyn/issues/23969 -->
```

Comment:
@jaredpar Can these 3 issues with PathMap be addressed in 15.7 so we can clean this up?

Reply:
@jaredpar Ping.

Reply:
Is this asking to add a new warning / error to the compiler in the default config?

Reply:
Yes, that's one of the issues. Right now if you pass `/deterministic` and `/pathmap` that doesn't cover all paths you don't know that the compiler outputs are still non-deterministic.

Reply:
This particular issue is not blocking for the SourceLink change. It's the other issues with quoting and escaping that are.

Reply:
We cannot add a new diagnostic to an existing path without user opt-in. That's pretty much a non-starter. 

---
In reply to: [175848556](https://github.com/dotnet/roslyn/pull/25325#discussion_r175848556) [](ancestors = 175848556)

Reply:
Can we do it with `/strict`?

Reply:
No. That's not a tool we ask users to rely on.

Reply:
It's not that I don't see value in the switch, I just understand the constraints we live under. There are scenarios which are far more valuable and easy to detect for which I can't issue new diagnostics.

---
In reply to: [176509304](https://github.com/dotnet/roslyn/pull/25325#discussion_r176509304) [](ancestors = 176509304)

Reply:
This scenario is somewhat unique though since one can't write an analyzer that reports these errors currently. We would need to expose a new API that the analyzer could hook into. This would be a callback that's invoked on every path that's embedded into the DLL/PDB. Then an analyzer could hook that and report errors if the paths do not match the PathMap. Would that be a preferred way?

Reply:
I guess the analyzer can be written today that finds all the paths that end up embedded into the DLL/PDB, but it'd be hard to keep it up to date with the compiler.